File: go/spore-core/tools/fixture_replay_test.go
// Fixture-replay tests for issue #5.
//
// Loads shared fixtures from /fixtures/tools and asserts the Go
// implementation produces the same outcome categories the Rust, TypeScript,
// and Python implementations do.

package tools

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// fixturesDir locates /fixtures/tools relative to this test file.
func fixturesDir(t *testing.T) string {
	t.Helper()
	_, here, _, _ := runtime.Caller(0)
	// here = .../spore-core/go/spore-core/tools/fixture_replay_test.go
	// go up: tools → spore-core → go → repo root → fixtures/tools
	root := filepath.Join(filepath.Dir(here), "..", "..", "..", "fixtures", "tools")
	return root
}

// paramParseOK mirrors the Rust helper of the same name: returns true if
// parameter validation passes, false if InvalidParameters would be raised.
func paramParseOK(tool string, input json.RawMessage) bool {
	tryParse := func(v any) bool {
		if len(input) == 0 {
			return false
		}
		return json.Unmarshal(input, v) == nil
	}
	requireFields := func(v map[string]any, fields ...string) bool {
		var probe map[string]any
		if err := json.Unmarshal(input, &probe); err != nil {
			return false
		}
		for _, f := range fields {
			if _, ok := probe[f]; !ok {
				return false
			}
		}
		return true
	}
	_ = tryParse
	switch tool {
	case "read_file":
		return requireFields(nil, "path")
	case "write_file":
		return requireFields(nil, "path", "content")
	case "list_dir":
		return requireFields(nil, "path")
	case "delete_file":
		return requireFields(nil, "path")
	case "move_file":
		return requireFields(nil, "src", "dst")
	case "grep_files":
		if !requireFields(nil, "pattern", "path") {
			return false
		}
		var p GrepFilesParams
		if json.Unmarshal(input, &p) != nil {
			return false
		}
		_, err := regexp.Compile(p.Pattern)
		return err == nil
	case "find_files":
		return requireFields(nil, "glob", "path")
	case "git_status":
		return true
	case "git_log":
		var p GitLogParams
		return json.Unmarshal(input, &p) == nil
	case "git_diff":
		var p GitDiffParams
		return json.Unmarshal(input, &p) == nil
	case "git_commit":
		return requireFields(nil, "message")
	case "git_reset":
		return requireFields(nil, "target", "mode")
	case "http_get":
		return requireFields(nil, "url")
	case "http_post":
		return requireFields(nil, "url", "body")
	default:
		return true
	}
}

func TestFixtureReplayParamValidation(t *testing.T) {
	path := filepath.Join(fixturesDir(t), "param_validation.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Skipf("fixture not found: %v", err)
	}
	var scenarios []struct {
		Tool     string          `json:"tool"`
		Input    json.RawMessage `json:"input"`
		Expected string          `json:"expected"`
	}
	if err := json.Unmarshal(data, &scenarios); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if len(scenarios) == 0 {
		t.Fatal("expected >= 1 scenario")
	}
	for _, sc := range scenarios {
		actual := "invalid_parameters"
		if paramParseOK(sc.Tool, sc.Input) {
			actual = "ok"
		}
		if actual != sc.Expected {
			t.Errorf("scenario tool=%s got %s expected %s", sc.Tool, actual, sc.Expected)
		}
	}
}

func TestFixtureReplayOutputTruncation(t *testing.T) {
	path := filepath.Join(fixturesDir(t), "output_truncation.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Skipf("fixture not found: %v", err)
	}
	var scenarios []struct {
		ContentLength    int    `json:"content_length"`
		HeadTokens       uint32 `json:"head_tokens"`
		TailTokens       uint32 `json:"tail_tokens"`
		ExpectsTruncated bool   `json:"expects_truncated"`
	}
	if err := json.Unmarshal(data, &scenarios); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	sb := sporecore.DefaultSandbox{}
	for _, sc := range scenarios {
		content := strings.Repeat("x", sc.ContentLength)
		out := sb.HandleLargeOutput(context.Background(), content, "fx", sc.HeadTokens, sc.TailTokens)
		actuallyTruncated := out.Truncated
		if actuallyTruncated != sc.ExpectsTruncated {
			t.Errorf("content_length=%d head=%d tail=%d: truncated=%v, expected=%v",
				sc.ContentLength, sc.HeadTokens, sc.TailTokens, actuallyTruncated, sc.ExpectsTruncated)
		}
	}
}

func TestFixtureReplaySubagentScenarios(t *testing.T) {
	path := filepath.Join(fixturesDir(t), "subagent_scenarios.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Skipf("fixture not found: %v", err)
	}
	var scenarios []struct {
		Name           string `json:"name"`
		ChildRunResult struct {
			Kind   string `json:"kind"`
			Output string `json:"output"`
		} `json:"child_run_result"`
		ParentCallID string `json:"parent_call_id"`
		Expected     struct {
			Kind             string `json:"kind"`
			Content          string `json:"content"`
			Recoverable      *bool  `json:"recoverable"`
			ParentToolCallID string `json:"parent_tool_call_id"`
		} `json:"expected"`
	}
	if err := json.Unmarshal(data, &scenarios); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, sc := range scenarios {
		var runResult sporecore.RunResult
		switch sc.ChildRunResult.Kind {
		case "success":
			runResult = sporecore.RunResult{Kind: sporecore.RunSuccess, Output: sc.ChildRunResult.Output}
		case "failure":
			runResult = sporecore.RunResult{Kind: sporecore.RunFailure, Reason: sporecore.HaltReason{Kind: sporecore.HaltHumanHalted}}
		case "waiting_for_human":
			paused := &sporecore.PausedState{
				SessionID:    "s",
				Task:         sporecore.NewTask("x", "s", sporecore.LoopStrategy{Kind: sporecore.StrategyReAct, MaxIterations: 1}),
				HumanRequest: sporecore.HumanRequest{Kind: sporecore.HumanReqClarification, Question: "?"},
			}
			req := &sporecore.HumanRequest{Kind: sporecore.HumanReqClarification, Question: "?"}
			runResult = sporecore.RunResult{Kind: sporecore.RunWaitingForHuman, State: paused, Request: req}
		default:
			t.Fatalf("unknown child_run_result.kind %q", sc.ChildRunResult.Kind)
		}
		h := &scriptedHarness{results: []sporecore.RunResult{runResult}}
		child := sporecore.NewStandardToolRegistry()
		sub, err := NewSubagentTool("subagent", "child", json.RawMessage(`{"type":"object"}`),
			0, Isolated{}, h, child)
		if err != nil {
			t.Fatalf("[%s] new: %v", sc.Name, err)
		}
		callIn := sporecore.ToolCall{ID: sc.ParentCallID, Name: "subagent", Input: json.RawMessage(`{"instruction":"x"}`)}
		out := sub.Execute(context.Background(), callIn, sporecore.AllowAllSandbox{}, nil)

		switch sc.Expected.Kind {
		case "success":
			if out.Kind != sporecore.ToolOutputSuccess {
				t.Errorf("[%s] expected success, got %+v", sc.Name, out)
			}
			if sc.Expected.Content != "" && out.Content != sc.Expected.Content {
				t.Errorf("[%s] content mismatch: %q vs %q", sc.Name, out.Content, sc.Expected.Content)
			}
		case "error":
			if out.Kind != sporecore.ToolOutputError {
				t.Errorf("[%s] expected error, got %+v", sc.Name, out)
			}
			if sc.Expected.Recoverable != nil && out.Recoverable != *sc.Expected.Recoverable {
				t.Errorf("[%s] recoverable=%v, expected %v", sc.Name, out.Recoverable, *sc.Expected.Recoverable)
			}
		case "waiting_for_human":
			if out.Kind != sporecore.ToolOutputWaitingForHuman {
				t.Errorf("[%s] expected waiting_for_human, got %+v", sc.Name, out)
				continue
			}
			if out.ChildState == nil || out.ChildState.ParentToolCallID != sc.Expected.ParentToolCallID {
				t.Errorf("[%s] parent_tool_call_id mismatch: %+v vs %s",
					sc.Name, out.ChildState, sc.Expected.ParentToolCallID)
			}
		default:
			t.Fatalf("[%s] unknown expected.kind %q", sc.Name, sc.Expected.Kind)
		}
	}
}

File: typescript/packages/tools/tests/fixture-replay.test.ts
/**
 * Fixture-replay tests for the shared tool fixtures in `fixtures/tools/`.
 * Mirrors the Rust suite in `rust/crates/spore-core/src/tools/mod.rs`.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import {
  defaultHandleLargeOutput,
  DeleteFileParamsSchema,
  FindFilesParamsSchema,
  GitCommitParamsSchema,
  GitDiffParamsSchema,
  GitLogParamsSchema,
  GitResetParamsSchema,
  GitStatusParamsSchema,
  GrepFilesParamsSchema,
  HttpGetParamsSchema,
  HttpPostParamsSchema,
  ListDirParamsSchema,
  MoveFileParamsSchema,
  ReadFileParamsSchema,
  WriteFileParamsSchema,
} from "../src/index.js";

const here = dirname(fileURLToPath(import.meta.url));
const fixturesRoot = resolve(here, "../../../../fixtures/tools");

interface ParamValidationScenario {
  tool: string;
  input: unknown;
  expected: "ok" | "invalid_parameters";
}

function paramParseOk(tool: string, input: unknown): boolean {
  switch (tool) {
    case "read_file":
      return ReadFileParamsSchema.safeParse(input).success;
    case "write_file":
      return WriteFileParamsSchema.safeParse(input).success;
    case "list_dir":
      return ListDirParamsSchema.safeParse(input).success;
    case "delete_file":
      return DeleteFileParamsSchema.safeParse(input).success;
    case "move_file":
      return MoveFileParamsSchema.safeParse(input).success;
    case "grep_files": {
      const r = GrepFilesParamsSchema.safeParse(input);
      if (!r.success) return false;
      try {
        new RegExp(r.data.pattern);
      } catch {
        return false;
      }
      return true;
    }
    case "find_files":
      return FindFilesParamsSchema.safeParse(input).success;
    case "git_status":
      return GitStatusParamsSchema.safeParse(input).success;
    case "git_log":
      return GitLogParamsSchema.safeParse(input).success;
    case "git_diff":
      return GitDiffParamsSchema.safeParse(input).success;
    case "git_commit":
      return GitCommitParamsSchema.safeParse(input).success;
    case "git_reset":
      return GitResetParamsSchema.safeParse(input).success;
    case "http_get":
      return HttpGetParamsSchema.safeParse(input).success;
    case "http_post":
      return HttpPostParamsSchema.safeParse(input).success;
    default:
      return true;
  }
}

describe("fixture: param_validation", () => {
  const data = readFileSync(
    resolve(fixturesRoot, "param_validation.json"),
    "utf8",
  );
  const scenarios = JSON.parse(data) as ParamValidationScenario[];

  it("loads at least one scenario", () => {
    expect(scenarios.length).toBeGreaterThan(0);
  });

  for (const sc of scenarios) {
    it(`${sc.tool} -> ${sc.expected}`, () => {
      const actual = paramParseOk(sc.tool, sc.input)
        ? "ok"
        : "invalid_parameters";
      expect(actual).toBe(sc.expected);
    });
  }
});

interface TruncationScenario {
  content_length: number;
  head_tokens: number;
  tail_tokens: number;
  expects_truncated: boolean;
}

describe("fixture: output_truncation", () => {
  const data = readFileSync(
    resolve(fixturesRoot, "output_truncation.json"),
    "utf8",
  );
  const scenarios = JSON.parse(data) as TruncationScenario[];

  for (const sc of scenarios) {
    it(`content_length=${sc.content_length} -> truncated=${sc.expects_truncated}`, async () => {
      const content = "x".repeat(sc.content_length);
      const out = await defaultHandleLargeOutput(
        content,
        "fx",
        sc.head_tokens,
        sc.tail_tokens,
      );
      expect(out.truncated).toBe(sc.expects_truncated);
    });
  }
});

interface SubagentScenario {
  name: string;
  child_run_result: {
    kind: "success" | "failure" | "waiting_for_human";
    output?: string;
  };
  parent_call_id: string;
  expected:
    | { kind: "success"; content: string }
    | { kind: "error"; recoverable: boolean }
    | { kind: "waiting_for_human"; parent_tool_call_id: string };
}

import {
  emptyAggregateUsage,
  emptyBudgetSnapshot,
  emptySessionState,
  harnessTesting,
  newTask,
  SessionId,
  TaskId,
  toolRegistry,
  type Harness,
  type HarnessRunOptions,
  type HumanResponse,
  type PausedState,
  type RunResult,
  type StreamSink,
} from "@spore/core";
import { SubagentTool } from "../src/index.js";

const { AllowAllSandbox } = harnessTesting;
const { StandardToolRegistry } = toolRegistry;
// Storage seam (#75): SubagentTool ignores ctx, but the signature requires one.
const ctx = toolRegistry.toolRegistryMock.testCtx();

class ScriptedHarness implements Harness {
  constructor(private readonly r: RunResult) {}
  async run(_opts: HarnessRunOptions): Promise<RunResult> {
    return this.r;
  }
  async resume(
    _state: PausedState,
    _response: HumanResponse,
    _onStream?: StreamSink,
  ): Promise<RunResult> {
    return this.r;
  }
}

describe("fixture: subagent_scenarios", () => {
  const data = readFileSync(
    resolve(fixturesRoot, "subagent_scenarios.json"),
    "utf8",
  );
  const scenarios = JSON.parse(data) as SubagentScenario[];

  for (const sc of scenarios) {
    it(sc.name, async () => {
      let runResult: RunResult;
      const sid = SessionId.of("s");
      switch (sc.child_run_result.kind) {
        case "success":
          runResult = {
            kind: "success",
            output: sc.child_run_result.output ?? "done",
            session_id: sid,
            usage: emptyAggregateUsage(),
            turns: 1,
          };
          break;
        case "failure":
          runResult = {
            kind: "failure",
            reason: { kind: "human_halted" },
            session_id: sid,
            usage: emptyAggregateUsage(),
            turns: 1,
          };
          break;
        case "waiting_for_human":
          runResult = {
            kind: "waiting_for_human",
            state: {
              session_id: sid,
              task_id: TaskId.of("t"),
              turn_number: 1,
              session_state: emptySessionState(),
              pending_tool_calls: [],
              approved_results: [],
              human_request: { kind: "clarification", question: "?" },
              task: newTask("x", sid, {
                kind: "react",
                budget: { kind: "per_loop", value: 1 },
                agent: "",
                toolset: "",
              }),
              budget_used: emptyBudgetSnapshot(),
              child_state: null,
              toolset: "",
            },
            request: { kind: "clarification", question: "?" },
          };
          break;
      }

      const sub = SubagentTool.buildOrThrow({
        name: "subagent",
        description: "child",
        inputSchema: { type: "object" },
        timeoutMs: 5_000,
        contextSharing: { kind: "isolated" },
        harness: new ScriptedHarness(runResult),
        childRegistry: new StandardToolRegistry(),
      });

      const out = await sub.execute(
        {
          id: sc.parent_call_id,
          name: "subagent",
          input: { instruction: "x" },
        },
        new AllowAllSandbox(),
        ctx,
      );

      if (sc.expected.kind === "success") {
        expect(out.kind).toBe("success");
        if (out.kind !== "success") throw new Error("unreachable");
        expect(out.content).toBe(sc.expected.content);
      } else if (sc.expected.kind === "error") {
        expect(out.kind).toBe("error");
        if (out.kind !== "error") throw new Error("unreachable");
        expect(out.recoverable).toBe(sc.expected.recoverable);
      } else {
        expect(out.kind).toBe("waiting_for_human");
        if (out.kind !== "waiting_for_human") throw new Error("unreachable");
        expect(out.child_state.parent_tool_call_id).toBe(
          sc.expected.parent_tool_call_id,
        );
      }
    });
  }
});
